Small Reader Improvements: Topics Management

ParaskP7 opened the pull request:
Fixes #13972

This PR makes it possible for users to see the `Topic Picker` screen again after they have added some topics. Previously, this screen would only be shown once, within the `Discover` tab and only if the user hasn't followed any topics yet. With this change, even if the user has already followed some topics, they will still be able to see the `Topic Picker` screen through the `Manage Topics & Sites` settings screen.

Discover | Settings
-------|------
<img width="250" height="500" alt="discover" src="https://user-images.githubusercontent.com/9729923/109326115-6ebc6a00-785f-11eb-8686-6830515f9bf1.mp4"> | <img width="250" height="500" alt="settings" src="https://user-images.githubusercontent.com/9729923/109326516-f86c3780-785f-11eb-8e7e-974382b82906.mp4">

To test:
- Unfollow all topics and navigate to `Reader` -> `DISCOVER` tab.
- Make sure the `DISCOVER` screen is empty. Click on the `GET STARTED` button.
- Verify that the `Topic Picker` screen is shown and everything looks as expected. Nothing changes here. Notice the `Choose your interests` title and the `Select a few to continue` text on the disabled state of the button.
- Select a few topics and click the `Done` button.
- Verify that the `DISCOVER` screen is now showing some content.
- Find the ⚙️ icon on top, within the `Reader` toolbar, and click on it to navigate to the `Manage Topics & Sites` settings screen.
- Verify that at least a few topics are listed under the `FOLLOWED TOPICS` tab list.
- Verify that the `cross` icon has been replaced with the `reader following` icon. Click on the `FAB` button.
- Verify that the `Topic Picker` screen is shown and everything looks familiar. Few things change here. Notice that the `Choose your interests` title is no longer shown. Instead the `Follow topics` toolbar title is shown. Also, notice that the `Select a few to continue` text on the disable state of the button has been now replaced with `Done`.
- Select a few topics and click the `Done` button.
- Verify that you get navigated back to the `Manage Topics & Sites` setting screen and that those newly picked topics are now shown within the `FOLLOWED TOPICS` tab list.
- Remove some topics by clicking the `reader following` icon and verify that this functionality works as expected.
- Go wild... play some more, select a lot of topics, select all the topics, verify everything works as expected, then remove lots of topics, remove all topics, check the `DISCOVER` tab again, make sure everything looks and feel okay, continue...

Note:
After talking to @mbshakti, we decided to add another improvement to the `Topics Picker` screen, where we will show some illustration and copy that there are no recommendations left on an empty state. But, this will be done on a separate PR when it will be design ready.

**Merge Instructions** (:warning:):
1. ~~Wait for @mbshakti to do design review.~~
1. ~~Remove the `Needs Design Review` label.~~
1. Merge this PR.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ParaskP7 committed:
Test: Reformat reader interests view model test

ParaskP7 committed:
Test: Make experimental coroutines api a class level annotation

ParaskP7 committed:
Test: Rename mock interests to plain interests

Having this naming convention creates a bit of confusion since those are
not mock interests but rather stub interests (fakes). Thus, this commit
removes the 'mock' keyword to increase cohesion.

ParaskP7 committed:
Refactor: Reformat reader interests view model

ParaskP7 committed:
Refactor: Reformat reader interests fragment

ParaskP7 committed:
Feature: Add entry point to reader interests view model

This new field will help the view model to be doing certain things
differently depending on the screen that launched and started this view
model (discover or settings).

ParaskP7 committed:
Refactor: Improve reader interests view model start function

This change does the following:
- Moves the 'isStart' logic to become the very first step.
- Removes the helper 'isLanguageSame()' function as it is being to basic
and instead uses the 'this.currentLanguage == currentLanguage' directly.
- Makes 'currentLanguage' the first parameter of the function since it
is used first in order.
- Moves the 'loadUserTags()' trigger to become the very last step.

ParaskP7 committed:
Refactor: Remove unnecessary r string import from view model

ParaskP7 committed:
Refactor: Remove redundant lambda arrow from view model

ParaskP7 committed:
Refactor: Remove redundant sam constructor from reader interest fragment

ParaskP7 committed:
Refactor: Extract check and load interests function on view model

ParaskP7 committed:
Test: Add tests for not closing reader on empty user tag for settings

These tests will fail as the implementation is not ready yet. The next
commit will make the tests pass (TDD).

ParaskP7 committed:
Feature: Implement not closing reader on empty user tag for settings

This commit makes the previous failing test pass (TDD).

ParaskP7 committed:
Test: Add tests for not showing excluded interests for settings

These tests will fail as the implementation is not ready yet. The next
commit will make the tests pass (TDD).

ParaskP7 committed:
Feature: Implement not showing excluded interests for settings

With this change all the interests that the user already picked up from
previous interaction with this screen, or manual adding them from
settings, will be excluded from the general list of interests next time
this screen is shown.

This commit makes the previous failing test pass (TDD).

ParaskP7 committed:
Test: Add tests for showing another disable button text for settings

These tests will fail as the implementation is not ready yet. The next
commit will make the tests pass (TDD).

ParaskP7 committed:
Feature: Implement showing another disable button text for settings

This commit makes the previous failing test pass (TDD).

ParaskP7 committed:
Refactor: Refactor title res related code in reader interest view model

ParaskP7 committed:
Test: Add tests for closing the reader interest for settings

These tests will fail as the implementation is not ready yet. The next
commit will make the tests pass (TDD).

ParaskP7 committed:
Feature: Implement closing the reader interest for settings

This commit makes the previous failing test pass (TDD).

ParaskP7 committed:
Test: Add tests for showing another disable button text for settings

This is now for the case when the screen has no interests at all.

These tests will fail as the implementation is not ready yet. The next
commit will make the tests pass (TDD).

ParaskP7 committed:
Feature: Implement showing another disable button text for settings

This is now for the case when the screen has no interests at all.

This commit makes the previous failing test pass (TDD).

ParaskP7 committed:
Test: Add tests for clicking on back button click in view model

These tests will fail as the implementation is not ready yet. The next
commit will make the tests pass (TDD).

ParaskP7 committed:
Feature: Implement clicking on back button click in view model

This commit makes the previous failing test pass (TDD).

ParaskP7 committed:
Test: Refactor init helper function on reader interest view model test

ParaskP7 committed:
Feature: Add reader interests activity to host reader interest fragment

This activity will be used from the settings screen in order to reuse
the 'ReaderInterestFragment' screen without further tuning.

ParaskP7 committed:
Feature: Launch reader interest screen from setting using a fab button

This solution connects the settings screen with the reader interests
screen through the newly created fab button. In addition, this changes
makes sure that when the user returns from this screen an update is
performed so that the settings screen gets reloaded with the newly
picked interests (if any). Thus, activity for result is being utilised
during the launch.

ParaskP7 committed:
Feature: Get entry point from activity intent and pass it to view model

ParaskP7 committed:
Feature: Make parent view model optional when coming from settings

ParaskP7 committed:
Feature: Hide the fragment level back button when coming from settings

ParaskP7 committed:
Feature: Hide the fragment level title when coming from settings

ParaskP7 committed:
Feature: Close the parent activity when close is triggered from settings

ParaskP7 committed:
Feature: Change chip alignment to top from being shown in the center

This only becomes apparent when there are only few chips being displayed
on the screen, which happens when the users actually selects lots of
topics to follow.

ParaskP7 committed:
Feature: Change remove interest button from cross to reader following

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/reader/discover/interests/ReaderInterestsFragment.kt:
```diff
@@ -80,18 +86,28 @@ class ReaderInterestsFragment : Fragment(R.layout.reader_interests_fragment_layo
             with(uiHelpers) {
                 updateVisibility(progress_bar, uiState.progressBarVisible)
-                updateVisibility(title, uiState.titleVisible)
+                updateVisibility(title, if (entryPoint == EntryPoint.DISCOVER) uiState.titleVisible else false)
```
❓: Can we extract this title visibility if condition to view model?

ParaskP7 replied:
👋 @ashiagr !

You are totally right, on it.

ParaskP7 replied:
👋  @ashiagr !

This is now done: 9955d140007ec9a6cd529a7f95e6e7df082eac2e and 2db3918ea140b2dfb34ef41cf611812c90dbc348

ashiagr reviewed (commented):
Thanks for working on this @ParaskP7!

Everything works as described, awesome job with TDD 👏. I just left one question and noticed one issue when topics were selected after rotating the "Manage Topics and Sites" screen and were not displayed after clicking "Done" button.

I've captured a video for the issue, hope this helps.


https://user-images.githubusercontent.com/1405144/109460412-7d389a80-7a86-11eb-9b3a-4323b2b28e0c.mp4

ParaskP7 commented:
👋 @ashiagr !

> Everything works as described, awesome job with TDD 👏.

🙏 TDD rocks and works! 😃 

> ...and noticed one issue when topics were selected after rotating the "Manage Topics and Sites" screen and were not displayed after clicking "Done" button.

Really good find Ashita, many many thanks, will look at that asap. I have been rotating and doing stuff but never noticed that the items don't get to be updated after the first rotation, kudos to you! 🙏

ParaskP7 committed:
Test: Add tests for hiding the title for settings

These tests will fail as the implementation is not ready yet. The next
commit will make the tests pass (TDD).

ParaskP7 committed:
Feature: Implement hiding the title for settings

This commit makes the previous failing test pass (TDD).

ParaskP7 commented:
👋 @ashiagr !

> Really good find Ashita, many many thanks, will look at that asap. I have been rotating and doing stuff but never noticed that the items don't get to be updated after the first rotation, kudos to you! 🙏

After investigating this issue and as per our discussion, I opened a separate issue (see #14162) since this seems to be a more generic problem with the screen and not related to this newly added functionality.

ParaskP7 committed:
Analysis: Fix ktlint issue with kotlin 1.4 trailing commas

ashiagr reviewed (approved):
Thanks for this, all looks good to me 👍.

mbshakti commented:
Looks and words perfectly @ParaskP7 Great job!! 🙏 

I know you mentioned it's a separate issue but I had the same experience as ashita where if I followed a topic in landscape mode, it didn't add those to my followed list.

ParaskP7 commented:
> Looks and words perfectly @ParaskP7 Great job!! 🙏

❤️ 🙏 🌟 

> I know you mentioned it's a separate issue but I had the same experience as ashita where if I followed a topic in landscape mode, it didn't add those to my followed list.

Perfect, thanks for verifying that @mbshakti ! 🙏 

Just as an FYI, note the fact that this happens when you switch to landscape while within the `Manage Topics & Sites` settings screen. If you start the app (ie. tablet), or start with landscape before entering this screen, then you will not have this issue.

ParaskP7 commented:
PS: I also created this #14177 related issue for showing an illustration and copy on an empty state.
